File: Solutions/56-merge-intervals/merge-intervals.swift
class Solution {
    func merge(_ intervals: [[Int]]) -> [[Int]] {
        
        // Step 1: Return empty if input is empty
        guard !intervals.isEmpty else { return [] }
        
        // Step 2: Sort intervals by start time
        let sortedIntervals = intervals.sorted { $0[0] < $1[0] }
        
        // Step 3: Initialize merged array
        var merged: [[Int]] = []
        
        // Step 4: Iterate through each interval
        for interval in sortedIntervals {
            
            // Step 5: If merged is empty, just append the interval
            if merged.isEmpty {
                merged.append(interval)
            } else {
                // Step 6: Get the last interval in merged
                var last = merged.removeLast()
                
                // Step 7: Check for overlap
                if interval[0] <= last[1] {
                    // Merge intervals: update the end to max of both
                    last[1] = max(last[1], interval[1])
                    merged.append(last)
                } else {
                    // No overlap, append both intervals
                    merged.append(last)
                    merged.append(interval)
                }
            }
        }
        
        // Step 8: Return the merged intervals
        return merged
    }
}

File: Solutions/56-merge-intervals/merge-intervals.scala
object Solution {
    def merge(intervals: Array[Array[Int]]): Array[Array[Int]] = {

        // Edge case: if there are 0 or 1 intervals, nothing to merge
        if (intervals.length <= 1) return intervals

        // Step 1: Sort intervals based on the starting time
        // This ensures that overlapping intervals come next to each other
        val sorted = intervals.sortBy(interval => interval(0))

        // Step 2: Create a list to store merged intervals
        val merged = scala.collection.mutable.ArrayBuffer[Array[Int]]()

        // Step 3: Initialize with the first interval
        var currentStart = sorted(0)(0)
        var currentEnd = sorted(0)(1)

        // Step 4: Iterate through the rest of the intervals
        for (i <- 1 until sorted.length) {

            val start = sorted(i)(0)
            val end = sorted(i)(1)

            // Case 1: Overlapping intervals
            // If the current interval's start is <= currentEnd,
            // it overlaps with the previous one
            if (start <= currentEnd) {
                // Merge by extending the end if needed
                currentEnd = Math.max(currentEnd, end)
            } else {
                // Case 2: No overlap
                // Save the previous interval
                merged += Array(currentStart, currentEnd)

                // Start a new interval
                currentStart = start
                currentEnd = end
            }
        }

        // Step 5: Add the last interval
        merged += Array(currentStart, currentEnd)

        // Convert result to Array[Array[Int]] and return
        merged.toArray
    }
}